Zend/zend_ini.c

```diff
@@ -560,33 +561,110 @@ static zend_ulong zend_ini_parse_quantity_internal(zend_string *value, zend_ini_
 
 	/* Ignore leading whitespace. ZEND_STRTOL() also skips leading whitespaces,
 	 * but we need the position of the first non-whitespace later. */
-	while (digits < str_end && zend_is_whitespace(*digits)) ++digits;
+	while (digits < str_end && zend_is_whitespace(*digits)) {++digits;}
 
 	/* Ignore trailing whitespace */
-	while (digits < str_end && zend_is_whitespace(*(str_end-1))) --str_end;
+	while (digits < str_end && zend_is_whitespace(*(str_end-1))) {--str_end;}
 
 	if (digits == str_end) {
 		*errstr = NULL;
 		return 0;
 	}
 
-	zend_ulong retval;
-	errno = 0;
+	bool is_negative = false;
+	if (digits[0] == '+') {
+		++digits;
+	} else if (digits[0] == '-') {
+		is_negative = true;
+		++digits;
+	}
 
-	if (signed_result == ZEND_INI_PARSE_QUANTITY_SIGNED) {
-		retval = (zend_ulong) ZEND_STRTOL(digits, &digits_end, 0);
-	} else {
-		retval = ZEND_STRTOUL(digits, &digits_end, 0);
+	/* if there is no digit after +/- */
+	if (!isdigit(digits[0])) {
+		/* Escape the string to avoid null bytes and to make non-printable chars
+		 * visible */
+		smart_str_append_escaped(&invalid, ZSTR_VAL(value), ZSTR_LEN(value));
+		smart_str_0(&invalid);
+
+		*errstr = zend_strpprintf(0, "Invalid quantity \"%s\": no valid leading digits, interpreting as \"0\" for backwards compatibility",
+						ZSTR_VAL(invalid.s));
+
+		smart_str_free(&invalid);
+		return 0;
+	}
+
+	int base = 0;
+	if (digits[0] == '0' && !isdigit(digits[1])) {
+		/* Value is just 0 */
+		if ((digits+1) == str_end) {
+			*errstr = NULL;
+			return 0;
+		}
+
+		switch (digits[1]) {
+			/* Multiplier suffixes */
+			case 'g':
+			case 'G':
+			case 'm':
+			case 'M':
+			case 'k':
+			case 'K':
+				goto evaluation;
+			case 'x':
+			case 'X':
+				base = 16;
+				break;
```

Comment:
Nit: We could simplify this to

``` suggestion
	if (digits[0] == '0') {
		switch (digits[1]) {
			case 'x':
			case 'X':
				base = 16;
				digits += 2;
				break;
```

This allows to remove the special case for `0`, and the multiplier suffix cases in the switch.

We can then remove the `if (UNEXPECTED(digits == str_end)) {` block (we can add a special case for "no digits after base prefix" in the "no valid leading digits" bellow when base is not 0.)

Reply:
I've tried doing this and if I try to simplify it then I don't think I can have an error for invalid prefixes